### ISSUE TITLE
Paths with URL encoded characters

### ISSUE THREAD

**Issue by Tenzer** [thumbor/thumbor-aws#66]
Having just migrated from Thumbor 6 and tc-aws to Thumbor 7 and thumbor-aws, I found we had an issue with requests for files with names that require URL encoded characters. They would all result in 404s when using the thumbor-aws loader.

I narrowed this down to be a missing URL decode call on the file path before loading the file from S3. You can see the equivalent decode call in tc-aws here: https://github.com/thumbor-community/aws/blob/master/tc_aws/loaders/__init__.py#L20.

I tried a small fix by creating a new loader that did the URL decoding and otherwise passed the requests onto `thumbor_aws.loader`:
```py
from urllib.parse import unquote

from thumbor_aws.loader import load as upstream_load


async def load(context, path):
    """Patched version of thumbor_aws.loader.load which does URL decoding before requesting file from S3."""

    path = unquote(path)
    return await upstream_load(context, path)
```

That fixed the problem for me.

I then tried to see if I could contribute a fix for this in thumbor-aws, but I haven't been able to reproduce the problem in tests, which makes me wonder if there's something about our setup that means we exhibit this problem.

We've got requests arriving via CloudFront -> Nginx (does remapping of URLs) -> Thumbor. I do however not see how URLs for filenames that for instance contain spaces otherwise could be passed around, unless they are URL encoded, and thus would need to be URL decoded before requesting them from S3.

Adding something like my workaround to the `thumbor_aws.loader` doesn't break any of the tests, which makes me think it might be okay to add it in, but I would also like to be able to reproduce the problem in tests first.

Does paths that require URL encoded characters work fine for other people?

**Comment by Tenzer** [comment]
Please don't close this issue as being stale. It's not an issue that solves itself by not doing anything.

**Comment by aaugustin** [comment]
I am hitting the same issue.

To debug it, I tried to follow where Thumbor is manipulating URL-encoded vs. URL-decoded URLs. Unfortunately, the logic is quite confusing. It looks like some `quote` and `unquote` call were added in random places to address various problems, without a clear plan. This ends up with loaders trying to magically guess whether they're handed a URL-encoded or URL-decoded URL:

https://github.com/thumbor/thumbor/blob/3601a01f63ab4eba159abbd8770d236163b0fe16/thumbor/loaders/http_loader.py#L30-L34

It looks to me like `normalize_url` in `loaders.py`, `Storage.normalize_path` in `storage.py`, and `Storage.normalize_path` in `result_storage.py` should all apply the same logic i.e. unquote the path.

(Also, the first one was fixed for the case when the prefix is empty in 0.4.1, while the others still have the bug today.)
